### ISSUE TITLE
Expanding in a 'module-begin context can cause spurious use-before-definition errors

### ISSUE THREAD

**Issue by lexi-lambda** [racket/racket#2150]
This program is a little evil, but I *think* it should work:

```racket
#lang racket/base

(require (for-syntax racket/base
                     syntax/strip-context))

(begin-for-syntax
  (local-expand
   (syntax-local-lift-require
    'racket/base
    (strip-context
     #'(#%module-begin
        (require (for-syntax racket/base))

        (define-syntax (m stx)
          (f #'(void)))

        (define-for-syntax (f x) x)

        (m))))
   'module-begin
   '()))
```

However, it produces an error:

```
f: undefined;
 cannot reference an identifier before its definition
```

I think this is wrong, since I think this program should run successfully. By the time `(m)` is expanded, `f` should be initialized, but it apparently isn’t. Moving the definition of `f` above the definition of `m` makes the problem go away, but I don’t think it should matter.

**Comment by lexi-lambda** [comment]
Interestingly, I can reproduce this issue without `syntax-local-lift-require` if I perform local expansion inside a macro transformer instead of within `begin-for-syntax`. Here’s a simpler test case:

```racket
#lang racket/base

(require (for-syntax racket/base
                     syntax/strip-context))

(define-syntax (go stx)
  (local-expand
   #'(#%plain-module-begin
      (require (for-syntax racket/base))

      (define-syntax (m stx)
        (f #'(void)))

      (define-for-syntax (f x) x)

      (m))
   'module-begin
   '())
  #'(void))

(go)
```

This program also runs on Racket 6.12 (unlike the original program, which fails with a different, unrelated error on versions other than Racket 7), and it produces the same error that the Racket 7 expander does.

**Comment by mflatt** [comment]
I think the macro expander just can't do this. A reference syntactically before definition at phase > 0 works only as long as the defined identifier doesn't have any extra scopes relative to the enclosing module.

The problem is that the expander has to compile the implementation of `m` before it proceeds to expand the next form in the body (because that next form might use `m`). At that point, `f` isn't bound — so the expander has to guess at what the binding will be; specifically, it has to guess at a symbolic name used to represent the binding for `f`. It guesses the symbol in the identifier, a plain `'f`, and that guess is only right if the defined identifier `f` ends up with no extra scopes.

Guessing `'f` allows many patterns to work, such as when a macro is used to introduce a reference to a compile-time `f` that has not yet been bound at at phase 1, but will be defined through a non-macro-introduced definition.

I have never found a workable alternative to this compromise. To put it all another way, phase > 0 within a module is distressingly like the top level, because it has to make certain decisions before seeing the whole module.

**Comment by lexi-lambda** [comment]
Interesting. What you say makes sense, but it definitely isn’t an answer I anticipated. It’s a bit troubling, though, since it does lie directly in the way of what I was attempting to do, which is a sign my bundle of hacks may be unworkable in general in a way I did not expect.

Let me give you a little more context by email—I don’t want to put all of the gory details into this GitHub issue, but it’s possible you’ll have a better approach than what I’m currently trying to do.
